pandas/io/excel.py

```diff
@@ -1610,35 +1672,57 @@ def _convert_to_style(self, style_dict, num_format_str=None):
             return None
 
         # Create a XlsxWriter format object.
-        xl_format = self.book.add_format()
+        props = {}
 
         if num_format_str is not None:
-            xl_format.set_num_format(num_format_str)
+            props['num_format'] = num_format_str
 
```

Comment:
Do you mean explicitly moving the number format logic out? Yes, perhaps that's a worthwhile refactoring. I think we should also be calculating the number format from the `display.precision` config. For which reason, I believe all of those changes belong in a different PR.

Or are you talking about moving this mapping logic out? Well currently we assume nested style dicts as an interchange format, which are well-suited to openpyxl but need conversion for all writers. The stuff in formats/ should remain relatively writer-agnostic.